### PR TITLE
bpo-43284: Update platform.win32_ver to use platform._syscmd_ver instead of sys.getwindowsversion().platform_version for determining accurate Windows version

### PR DESCRIPTION
**Pull request by shreyanavigyan**
`platform.win32_ver` derives the Windows version from `sys.getwindowsversion().platform_version` which in turn derives the version from `kernel32.dll` (which can be of a different version than Windows itself). Therefore this PR updates the `platform.win32_ver` to determine the version using the `platform` module's `_syscmd_ver` private function to return an accurate version.

More discussions are held at: https://bugs.python.org/issue43284

https://bugs.python.org/issue43284

### CODE REVIEW

File: Lib/platform.py [hunk 1]
```diff
@@ -365,7 +365,10 @@ def win32_ver(release='', version='', csd='', ptype=''):
         return release, version, csd, ptype
 
     winver = getwindowsversion()
-    maj, min, build = winver.platform_version or winver[:3]
+    cmdwinver = _syscmd_ver()[2].split('.')
+    for x in range(3):
+        cmdwinver[x] = int(cmdwinver[x])
+    maj, min, build = tuple(cmdwinver) or winver[:3]
```
Comment (eryksun) [inline Lib/platform.py]:
```suggestion
    try:
        maj, min, build = map(int, _syscmd_ver()[2].split('.'))
    except ValueError:
        maj, min, build = winver.platform_version or winver[:3]
```

`_syscmd_ver()` doesn't validate that the version string has at least 3 dot-delimited numbers. Also, if it fails, it returns an empty string. In the above suggested code, both cases will raise `ValueError`, either from calling `int` on an empty string, or from trying to unpack too many values. It's handled by falling back on the old code that uses `platform_version`. This situation could occur due to problems with creating the shell process, or if the shell's output format changes in a new release and no longer matches the regex.

Comment (shreyanavigyan) [reply Lib/platform.py]: Ok...let me commit your suggestion.

**Comment by zooba** [comment]
I also noticed that `_norm_version` has a bug that is back in the code path with this change:

```python
    try:
        ints = map(int, l)
    except ValueError:
        strings = l
    else:
        strings = list(map(str, ints))
```

The `ValueError` is never going to be raised there because `map` is lazy. We probably just want to merge the else block into the try: `strings = list(map(str, map(int, l)))`

**Comment by shreyanavigyan** [comment]
Are you suggesting to make the change in this PR itself?

**Comment by zooba** [comment]
Yes, please. (They don't exist, but any tests for these cases would be breaking on your PR because you've reintroduced the call to that function.)

**Comment by shreyanavigyan** [comment]
Change applied.

File: Lib/platform.py [hunk 2]
```diff
@@ -238,12 +238,7 @@ def _norm_version(version, build=''):
     l = version.split('.')
     if build:
         l.append(build)
-    try:
-        ints = map(int, l)
-    except ValueError:
-        strings = l
-    else:
-        strings = list(map(str, ints))
+    strings = list(map(str, map(int, l)))
```
Comment (zooba) [inline Lib/platform.py]: We'll still need the `ValueError` check around this.

```suggestion
    try:
        strings = list(map(str, map(int, l)))
    except ValueError:
        strings = l
```

Comment (shreyanavigyan) [reply Lib/platform.py]: Forgot about that. Thanks for reminding.

**Comment by shreyanavigyan** [comment]
I also noticed something that can cause naming conflicts later. In the function win32_ver the minor version is stored in a variable named `min`  which is also the name of a built-in function. Therefore I propose that why not change the variable name `min` to `minor` to avoid naming conflicts. I also think the `maj` variable should be changed to `major`.

**Comment by zooba** [comment]
If you'd like to make more changes, go ahead. Just let us know when it's ready.

**Comment by shreyanavigyan** [comment]
Ok

**Comment by shreyanavigyan** [comment]
PR is ready for reviewing and if applicable then merging. (No more commits will be made)

**Comment by szmcdull** [comment]
calling shell command may lead to security vulnerability.  I'm strongly against this. I'm implementing a cpython wrapper that would disable methods like `os.system`, `subprocess.Popen` etc. But I found that many packages are calling platform.system(). 

IMO https://bugs.python.org/issue43284 is the expected behavior. The process is running in compatible mode, so an old version was reported. If you want the exact version, you should disable compatible mode.

**Comment by szmcdull** [comment]
In linux platform.system() also calls external command `uname` 😭

**Comment by eryksun** [comment]
> calling shell command may lead to security vulnerability. 

Running the `ver` command via `subprocess.check_output()` with `shell=True` executes the "%ComSpec%" shell. This should be the fully qualified path "%SystemRoot%\\System32\\cmd.exe". If `%ComSpec%` isn't a fully-qualified path, or if it's hijacked, then the system is already compromised.

Anyway, stating with Python 3.12 you'll probably be happy to discover that `win32_ver()` first tries a WMI query that's implemented by a new `_wmi` extension module. There are no plans to backport this extension module to previous versions, however.

> The process is running in compatible mode, so an old version was reported. If you want the exact version, you should disable compatible mode.

The platform module is intended to return information about the machine and installed operating system, for logging and such. On Windows, scripts should use `sys.getwindowsversion()` to get the API version. The latter can be a compatibility version. For example, if the interpreter is embedded in an application that has no manifest, the reported OS version is Windows 8.

**Comment by szmcdull** [comment]
> Running the ver command via subprocess.check_output() with shell=True executes the "%ComSpec%" shell. This should be the fully qualified path "%SystemRoot%\System32\cmd.exe". If %ComSpec% isn't a fully-qualified path, or if it's hijacked, then the system is already compromised.

Python should provide an option to disable external commands, just like PHP does. It's not only about hacking env variables, but also raises risks when string injection occurs. In my code I use C# to call python functions, which are user functions providing trading algorithms. But I don't want these user functions to call any external commands. Right now if I detect and prevent external command calls, packages like pandas/numpy would fail because they are calling platform.system(), which in turn calls external commands.

**Pull request merged**